File: wk6/java/backspaceStringCompare.java
// problem link: https://leetcode.com/problems/backspace-string-compare/

class Solution {
    public boolean backspaceCompare(String s, String t) {
        int ss = s.length(), tt = t.length(), sc = 0, tc = 0;
        while(ss >= 0 || tt >= 0) {
            while(--ss >= 0) {
                if(s.charAt(ss) == '#')
                    sc++;
                else if(sc > 0)
                    sc--;
                else
                    break;
            }
            while(--tt >= 0) {
                if(t.charAt(tt) == '#')
                    tc++;
                else if(tc > 0)
                    tc--;
                else
                    break;
            }
            if(ss >= 0 != tt >= 0)
                return false;
            if(ss < 0)
                break;
            if(s.charAt(ss) != t.charAt(tt))
                return false;
        }
        return true;
    }
}

File: wk6/cpp/backspaceStringCompare.cpp
// problem link: https://leetcode.com/problems/backspace-string-compare/description/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool backspaceCompare(string s, string t) {
        int sl = -1;
        for(int sr = 0; sr < s.length(); sr++) {
            if(s[sr] == '#')
                sl = max(-1, sl - 1);
            else
                s[++sl] = s[sr];
        }
        s.resize(sl + 1);
        int tl = -1;
        for(int tr = 0; tr < t.length(); tr++) {
            if(t[tr] == '#')
                tl = max(-1, tl - 1);
            else
                t[++tl] = t[tr];
        }
        t.resize(tl + 1);
        return s == t;
    }
};